Split locking

SamuelHall700 opened the pull request:
# CartograPhi Pull Request Template

Date: <!--- Include date PR was created -->   11/09/
Version Number: <!--- Include version number of CartograPhi the PR will be included in (e.g. 1.1.0 -> 1.1.1) -->   1.0.3 -> 1.1.1
 
## Description of change
Added new parameter 'split_lock' to the splitting conditions of a cellbox. If true when a cellbox is deemed homogenous, it will prevent it from splitting due to other splitting conditions. Default is false
<!--- Describe your changes in detail -->

## Fixes # (issue)
<!--- If this PR adds functionality or resolves problems associated with an issue on GitHub, please include a link to the issue -->
N/A

# Testing
To ensure that the functionality of the CartograPhi codebase remains consistent throughout the development cycle a testing strategy has been developed, which can be viewed in the document `test/testing_strategy.md`. 
This includes a collection of test files which should be run according to which part of the codebase has been altered in a pull request. Please consult the testing strategy to determine which tests need to be run. 

- [ ] My changes have not altered any of the files listed in the testing strategy

- [ ] My changes result in all required regression tests passing without the need to update test files.  
  
> *list which files have been altered and include a pytest.txt file for each of
> the tests required to be run*
>
> The files which have been changed during this PR can be listed using the command

    cartographi/__init__.py
    cartographi/cli.py
    cartographi/config_validation/config_validator.py
    cartographi/config_validation/mesh_schema.py
    cartographi/dataloaders/scalar/abstract_scalar.py
    cartographi/mesh_generation/boundary.py
    cartographi/mesh_generation/environment_mesh.py
    cartographi/mesh_generation/mesh_builder.py
    examples/environment_config/grf_example.config.json
    examples/environment_config/grf_minimal_example.config.json
    examples/environment_config/real_example.config.json
    tests/regression_tests/example_meshes/abstract_env_meshes/checkerboard_1.json
    tests/regression_tests/example_meshes/abstract_env_meshes/checkerboard_2.json
    tests/regression_tests/example_meshes/abstract_env_meshes/checkerboard_3.json
    tests/regression_tests/example_meshes/abstract_env_meshes/circle.json
    tests/regression_tests/example_meshes/abstract_env_meshes/circle_quadrant_nosplit.json
    tests/regression_tests/example_meshes/abstract_env_meshes/circle_quadrant_split.json
    tests/regression_tests/example_meshes/abstract_env_meshes/hgrad.json
    tests/regression_tests/example_meshes/abstract_env_meshes/vgrad.json
    tests/regression_tests/example_meshes/env_meshes/grf_downsample.json
    tests/regression_tests/example_meshes/env_meshes/grf_normal.json
    tests/regression_tests/example_meshes/env_meshes/grf_reprojection.json
    tests/regression_tests/example_meshes/env_meshes/grf_sparse.json
    tests/regression_tests/test_mesh.py
    tests/unit_tests/resources/feb_2013_Jgrid_config.json
    tests/unit_tests/resources/global_grf_normal.json

- [x] My changes require one or more test files to be updated for all regression tests to pass.   

All regression tests pass but require the new parameter to be added to the config sections of the test files. 
[pytest_test_mest.txt](https://github.com/antarctica/CartograPhi/files/12574875/pytest_test_mest.txt)

# Checklist

- [x] My code follows [pep8](https://peps.python.org/pep-0008/) style guidelines.  
- [x] I have commented my code, particularly in hard-to-understand areas.  
- [x] I have updated the documentation of the codebase where required.  
- [x] My changes generate no new warnings.   
- [ ] My PR has been made to the `1.1.x` branch (**DO NOT SUBMIT A PR TO MAIN**)  

**I'm submitting this to main due to merge conflicts with the 1.1.x branch _(some changes have been submitted to main that aren't in the 1.1.x branch yet)_. I'll create a new PR to the 1.1.x branch to pull from main after this.**

ayat-khairy reviewed (approved):
I second Harry's Key error point. Otherwise looks good.

SamuelHall700 commented on cartographi/dataloaders/scalar/abstract_scalar.py:
`if splitting_conds['split_lock']` will resolve to `false` if there isn't a value in the config file, hence why the default is false. This probably isn't best practice, so i'll update it to make it more clear

SamuelHall700 commented:
I've changed the implementation of the `split_lock` functionality to be clearer. 

The docstring of `get_hom_condition(...)` has been updated to include the `split_lock` parameter, so it should automatically be added to the docs on rebuild. 

The get_hom_condition of vector dataloaders never resolves to HOM, so there is no clear way of extending this change to the vector dataloaders.

I've replaced MIN with CLR in the abstract vector and abstract scalar dataloaders.

SamuelHall700 commented:
> Fair points. Can you add logging to show when `MIN` would have normally triggered? We should include that if just for our own debugging purposes in the future. After that, happy to approve the request

Added additional logging for minimum datapoints (entirely written by co-pilot, I guess the LLM thinks it's a good idea too!)
